Fix Passwordless default Cloud on first user signup

mcbattirola opened the pull request:
This PR addresses a bug in Cloud clusters where the first user signup with Passwordless does not update the cluster auth preferences to passwordless as expected.

This was working at some point, but the logic stopped working because it was checking for the number of existing users in the cluster:

```go
// Only set the connector name on the first user registration.
if len(users) != 1 {
	return nil
}
```

With the introduction of a preset user, this condition fails -- by the time the first user registers or updates their auth method, len(users) == 2.

### Fix


This PR fixes this bug by updating this logic to check if exactly 1 non-preset user exists, instead of checking the total amount of users.

I've added a function `getPresetUsers` to contain all preset users, following the pattern of `GetPresetRoles` in this same file. This way, this code should still work if more preset users are added.

`hasOneNonPresetUser` also does some extra work to make sure we account for deleted preset users (although this is not possible today).

Closes https://github.com/gravitational/cloud/issues/9002

mcbattirola committed:
Update condition to set passwordless to acount for preset users

mcbattirola committed:
Merge branch 'master' into mcbattirola/passwordless-default-cloud

mcbattirola committed:
Improve method

mcbattirola committed:
Merge branch 'master' of github.com:gravitational/teleport into mcbattirola/passwordless-default-cloud

mcbattirola committed:
Add test

mcbattirola committed:
Merge branch 'master' into mcbattirola/passwordless-default-cloud

mcbattirola committed:
Merge branch 'master' into mcbattirola/passwordless-default-cloud

mcbattirola committed:
Merge branch 'master' into mcbattirola/passwordless-default-cloud

codingllama commented on lib/auth/auth_with_roles.go:
```diff
@@ -3479,8 +3479,7 @@ func (a *ServerWithRoles) trySettingConnectorNameToPasswordless(ctx context.Cont
 
-	// Only set the connector name on the first user registration.
-	if len(users) != 1 {
+	if !hasOneNonPresetUser(users) {
```
Add a comment explaining why having exactly one non-preset user important, or maybe restore the previous comment?

mcbattirola replied:
Yes, that's a good idea, ty!

codingllama commented on lib/auth/init.go:
```diff
@@ -1048,12 +1048,26 @@ type PresetUsers interface {
+		presetUsers = append(presetUsers, accessBotUser)
+	}
+
+	// Certain `New$FooUser()` functions will return a nil role if the
+	// corresponding feature is disabled. They should be filtered out as they
+	// are not actually made available on the cluster.
+	return slices.DeleteFunc(presetUsers, func(r types.User) bool { return r == nil })
```
I'm not sure I follow this block. Isn't the `if accessBotUser != nil` above enough to handle the corner cases?

mcbattirola replied:
Yes, you're right, thanks.
I added the `DeleteFunc` to mimic the pattern of `getPresetRoles`, but it wasn't necessary the way I implemented `getPresetUsers`:

```go
// GetPresetRoles returns a list of all preset roles expected to be available on
// this cluster.
func GetPresetRoles() []types.Role {
	presets := []types.Role{
		services.NewPresetGroupAccessRole(),
		// ...
	}

	// Certain `New$FooRole()` functions will return a nil role if the
	// corresponding feature is disabled. They should be filtered out as they
	// are not actually made available on the cluster.
	return slices.DeleteFunc(presets, func(r types.Role) bool { return r == nil })
}
```

I've updated `getPresetUsers` to follow the same pattern, so `slices.DeleteFunc` is useful. What do you think?

```go
// getPresetUsers returns a list of all users roles expected to be available on
// this cluster.
func getPresetUsers() []types.User {
	presets := []types.User{
		services.NewSystemAutomaticAccessBotUser(),
	}

	// Certain `New$FooUser()` functions will return a nil role if the
	// corresponding feature is disabled. They should be filtered out as they
	// are not actually made available on the cluster.
	return slices.DeleteFunc(presets, func(r types.User) bool { return r == nil })
}
```

codingllama replied:
Got it. Makes sense.

codingllama commented on lib/auth/init_test.go:
```diff
@@ -906,6 +906,20 @@ func TestPresets(t *testing.T) {
+	modules.SetTestModules(t, &modules.TestModules{
+		TestBuildType: modules.BuildOSS,
+	})
+	require.Empty(t, len(getPresetUsers()))
```
```suggestion
	require.Empty(t, getPresetUsers())
```

codingllama commented on lib/auth/init_test.go:
```diff
@@ -906,6 +906,20 @@ func TestPresets(t *testing.T) {
+	modules.SetTestModules(t, &modules.TestModules{
+		TestBuildType: modules.BuildEnterprise,
+	})
+	require.Len(t, getPresetUsers(), 1)
```
Should we check that this contains exactly the result of NewSystemAutomaticAccessBotUser, in both cases (OSS and Enterprise modules)?

mcbattirola replied:
For OSS we have to check for an empty slice, since `NewSystemAutomaticAccessBotUser` returns `nil`:

```go
// oss
require.Equal(t, []types.User{}, getPresetUsers())

// enterprise
require.Equal(t, []types.User{
	services.NewSystemAutomaticAccessBotUser(),
}, getPresetUsers())
```

What do you think? I like this better than the current len check.

codingllama replied:
I would do require.Empty for the first, but otherwise seems OK.

codingllama commented on lib/auth/auth_with_roles_test.go:
```diff
@@ -8398,3 +8398,185 @@ func TestIsMFARequired_AdminAction(t *testing.T) {
+			// create preexisting users
+			for i := 0; i < tc.qtyPreexistingUsers; i += 1 {
+				_, _, err = CreateUserAndRole(srv.AuthServer, fmt.Sprintf("testuser-%d", i), nil, []types.Rule{
+					types.NewRule(types.KindUser, services.RW()),
```
Why is the rule needed?

mcbattirola replied:
Not necessary, thanks.

codingllama commented on lib/auth/auth_with_roles_test.go:
```diff
@@ -8398,3 +8398,185 @@ func TestIsMFARequired_AdminAction(t *testing.T) {
+				require.NoError(t, err)
+			}
+
+			// add the user that will have the auth method changed to passwordless
```
```suggestion
			// add the user that might have the auth method changed to passwordless
```

codingllama commented on lib/auth/auth_with_roles_test.go:
```diff
@@ -8398,3 +8398,185 @@ func TestIsMFARequired_AdminAction(t *testing.T) {
+			_, registerSolved, err := NewTestDeviceFromChallenge(registerChal, WithPasswordless())
+			require.NoError(t, err)
+
+			setupAuthContext, err := srv.Authorizer.Authorize(authz.ContextWithUser(ctx, TestAdmin().I))
+			require.NoError(t, err)
+
+			// actual authentication method change via ServerWithRoles
+			serverWithRoles := &ServerWithRoles{
+				authServer: srv.AuthServer,
+				alog:       srv.AuditLog,
+				context:    *setupAuthContext,
+			}
+
+			serverWithRoles.ChangeUserAuthentication(ctx, &proto.ChangeUserAuthenticationRequest{
+				TokenID:                token,
+				NewMFARegisterResponse: registerSolved,
+			})
```
Could we go through the gRPC server instead of reaching out to ServerWithRoles ourselves, possibly with a proxy-like identity?

Same for the other methods.

codingllama commented on lib/auth/auth_with_roles_test.go:
```diff
@@ -8398,3 +8398,185 @@ func TestIsMFARequired_AdminAction(t *testing.T) {
+		t.Run(tc.name, func(t *testing.T) {
+			// append preset to users list
+			users := append(tc.users, tc.existingPresets...)
+			actual := hasOneNonPresetUser(users)
```
I don't think we need to test the private method directly, this should already be covered by TestCloudDefaultPasswordless.

codingllama commented on lib/auth/auth_with_roles_test.go:
```diff
@@ -8398,3 +8398,185 @@ func TestIsMFARequired_AdminAction(t *testing.T) {
 	}
 }
+
+func TestCloudDefaultPasswordless(t *testing.T) {
```
Did you check for a test for the old `len(users)==1` logic, just in case?

mcbattirola replied:
Yeah, I actually found out that the test servers don't create the preset users, so I created them in the tests to simulate the runtime behavior. Then I added one test that catches the original bug (first user when presets exist) to make sure we're solving it.

TY!

mcbattirola committed:
Improve comment

Co-authored-by: Alan Parra <alan.parra@goteleport.com>

mcbattirola committed:
Apply code review comments

- Add comment back
- Fix test require.Empty condition
- Remove unnecessary test

mcbattirola committed:
Update getPresetUsers implementation to mimic getPresetRole pattern

codingllama commented on lib/auth/init.go:
```diff
@@ -1048,12 +1048,23 @@ type PresetUsers interface {
+	// Certain `New$FooUser()` functions will return a nil role if the
+	// corresponding feature is disabled. They should be filtered out as they
+	// are not actually made available on the cluster.
+	return slices.DeleteFunc(presets, func(r types.User) bool { return r == nil })
```
Maybe? `r` seems out of nowhere.

```suggestion
	return slices.DeleteFunc(presets, func(u types.User) bool { return u == nil })
```

mcbattirola committed:
Remvoe unnecessary rules on test roles

mcbattirola committed:
Update arg name in slices.DeleteFunc

Co-authored-by: Alan Parra <alan.parra@goteleport.com>

mcbattirola committed:
Assert expected values in TestGetPresetUsers

codingllama commented on lib/auth/auth_with_roles_test.go:
```diff
@@ -8454,17 +8454,13 @@ func TestCloudDefaultPasswordless(t *testing.T) {
-				_, _, err = CreateUserAndRole(srv.AuthServer, fmt.Sprintf("testuser-%d", i), nil, []types.Rule{
-					types.NewRule(types.KindUser, services.RW()),
-				})
+				_, _, err = CreateUserAndRole(srv.AuthServer, fmt.Sprintf("testuser-%d", i), nil, []types.Rule{})
```
You could do:

```go
				_, _, err = CreateUserAndRole(srv.AuthServer, fmt.Sprintf("testuser-%d", i), nil, nil)
```

personally I like to sneak in a few comments on ambiguous types for clarity:

```suggestion
				_, _, err = CreateUserAndRole(srv.AuthServer, fmt.Sprintf("testuser-%d", i), nil /* allowedLogins */, nil /* allowRules */)
```

codingllama commented on lib/auth/auth_with_roles_test.go:
```diff
@@ -8454,17 +8454,13 @@ func TestCloudDefaultPasswordless(t *testing.T) {
-			_, _, err = CreateUserAndRole(srv.AuthServer, username, nil, []types.Rule{
-				types.NewRule(types.KindUser, services.RW()),
-			})
+			_, _, err = CreateUserAndRole(srv.AuthServer, username, nil, []types.Rule{})
```
```suggestion
			_, _, err = CreateUserAndRole(srv.AuthServer, username, nil /* allowedLogins */, nil /* allowRules */)
```

mcbattirola committed:
Apply suggestions from code review - Improve tests

Co-authored-by: Alan Parra <alan.parra@goteleport.com>

mcbattirola committed:
Merge branch 'mcbattirola/passwordless-default-cloud' of github.com:gravitational/teleport into mcbattirola/passwordless-default-cloud

mcbattirola committed:
TestCloudDefaultPasswordless: use gRPC client

codingllama commented on lib/auth/auth_with_roles_test.go:
```diff
@@ -8449,28 +8448,50 @@ func TestCloudDefaultPasswordless(t *testing.T) {
+			// add permission to edit users, necessary to change auth method to passwordless
+			user, err := types.NewUser(u.username)
+			require.NoError(t, err)
+			role := services.RoleForUser(user)
```
Create a different role, or fetch the existing user role and modify it?

mcbattirola replied:
I went with the fetch and modify route, see https://github.com/gravitational/teleport/pull/41956#discussion_r1718990131. TY.

codingllama commented on lib/auth/auth_with_roles_test.go:
```diff
@@ -8449,28 +8448,50 @@ func TestCloudDefaultPasswordless(t *testing.T) {
+			upsertedRole, err := srv.Auth().UpsertRole(ctx, role)
+			require.NoError(t, err)
+			user.AddRole(upsertedRole.GetName())
+			srv.Auth().UpdateUser(ctx, &types.UserV2{})
```
This line looks like a leftover.

```suggestion```

codingllama commented on lib/auth/auth_with_roles_test.go:
```diff
@@ -8449,28 +8448,50 @@ func TestCloudDefaultPasswordless(t *testing.T) {
+			})
+			upsertedRole, err := srv.Auth().UpsertRole(ctx, role)
+			require.NoError(t, err)
+			user.AddRole(upsertedRole.GetName())
+			srv.Auth().UpdateUser(ctx, &types.UserV2{})
+			_, err = srv.Auth().UpsertUser(ctx, user)
+			require.NoError(t, err)
```
If we are using the "user role", isn't it already assigned to the user by some of the helpers above?

mcbattirola replied:
You're right, `createUserWithSecondFactors` calls `CreateUserAndRole` internally, which in turn creates a role for the user already, thanks. Updated in https://github.com/gravitational/teleport/pull/41956/commits/8248aec5ac417aff44e99e4747e77df072e4aff2

codingllama commented on lib/auth/auth_with_roles_test.go:
```diff
@@ -8449,28 +8448,50 @@ func TestCloudDefaultPasswordless(t *testing.T) {
+			client, err := srv.NewClient(TestUser(user.GetName()))
 			require.NoError(t, err)
 
+			client.SetMFAPromptConstructor(func(po ...mfa.PromptOpt) mfa.Prompt {
+				return mfa.PromptFunc(func(ctx context.Context, chal *proto.MFAAuthenticateChallenge) (*proto.MFAAuthenticateResponse, error) {
+					return u.webDev.SolveAuthn(chal)
+				})
+			})
```
Is this needed if we are solving the challenges explicitly below?

mcbattirola replied:
No, thank you!

codingllama commented on lib/auth/auth_with_roles_test.go:
```diff
@@ -8480,22 +8501,12 @@ func TestCloudDefaultPasswordless(t *testing.T) {
 			_, registerSolved, err := NewTestDeviceFromChallenge(registerChal, WithPasswordless())
 			require.NoError(t, err)
```
Use `u.webDev` instead of registering another device?

codingllama commented on lib/auth/auth_with_roles_test.go:
```diff
@@ -8480,22 +8501,12 @@ func TestCloudDefaultPasswordless(t *testing.T) {
-			}
-
-			serverWithRoles.ChangeUserAuthentication(ctx, &proto.ChangeUserAuthenticationRequest{
+			client.ChangeUserAuthentication(ctx, &proto.ChangeUserAuthenticationRequest{
```
Let's handle the resulting error too.

mcbattirola committed:
Remove blank UpdateUser and unnecessary MFA setup

mcbattirola committed:
Update existing role instead of creating a new one

codingllama commented on lib/auth/auth_with_roles_test.go:
```diff
@@ -8398,3 +8398,117 @@ func TestIsMFARequired_AdminAction(t *testing.T) {
+			role.SetRules(types.Allow, []types.Rule{
+				types.NewRule(types.KindUser, services.RW()),
+			})
+			upsertedRole, err := srv.Auth().UpsertRole(ctx, role)
+			require.NoError(t, err)
+			user.AddRole(upsertedRole.GetName())
+			_, err = srv.Auth().UpsertUser(ctx, user)
+			require.NoError(t, err)
```
The role is already assigned, I don't think you need the last update.

```suggestion
			_, err = srv.Auth().UpsertRole(ctx, role)
			require.NoError(t, err)
```

mcbattirola committed:
Make sure test server has preset users

mcbattirola committed:
Remove unnecessary user upsert after role change.

Co-authored-by: Alan Parra <alan.parra@goteleport.com>

flyinghermit commented on lib/auth/init.go:
```diff
@@ -1048,12 +1048,23 @@ type PresetUsers interface {
 	UpsertUser(ctx context.Context, user types.User) (types.User, error)
 }
 
+// getPresetUsers returns a list of all users roles expected to be available on
```
should be just users?
```suggestion
// getPresetUsers returns a list of all preset users expected to be available on
```

flyinghermit reviewed (approved):
LGTM

mcbattirola committed:
Improve `getPresetUsers` godoc

Co-authored-by: Sakshyam Shah <sshah@goteleport.com>

mcbattirola committed:
Merge branch 'master' into mcbattirola/passwordless-default-cloud
